File: frontend/src/utils/__tests__/test-bip322.ts
/**
 * Quick test script to verify BIP-322 signing matches Rust reference data
 * Run with: npx tsx frontend/src/utils/__tests__/test-bip322.ts
 */

import { Buffer } from 'buffer/';
import { sha256 } from 'js-sha256';
import { signMessage } from '../bitcoin-signer';

// Reference data from Rust test output
const testData = {
  privateKey: "b75a71ac262cdf072fc0b8dd813e1ef56ca77b957f7ae0a7998ce7f613b70351",
  publicKey: "1fbe8883da60fc9239cf2471c5a63fd3286bb1def2d3bb1b409d1ce82095da76",
  serializedMessage: "010000020000001fbe8883da60fc9239cf2471c5a63fd3286bb1def2d3bb1b409d1ce82095da7602020202020202020202020202020202020202020202020202020202020202020101010101010101010101010101010101010101010101010101010101010101010000000101000000000400000000010203",
  messageHash: "62313762326232643639373832623466323566633066646533623466316339396261656666366139373866363430346135336164616263363961333963383830",
  signature: "bc552cdbfd195014552400c63d4fe8d1e696f274f4a5294fae738dc96af2f7e86ec43b18f6e8076107104ec0691cbc9c4d870240325b5ed7713497e55e3029d0"
};

console.log('=== TypeScript BIP-322 Test ===\n');

// Test 1: Verify message hashing
console.log('Test 1: Message Hashing');
const serializedMessage = Buffer.from(testData.serializedMessage, 'hex');
console.log('  Serialized message length:', serializedMessage.length);

const firstHashHex = sha256(serializedMessage);
console.log('  First hash:', firstHashHex);

const firstHashBytes = Buffer.from(firstHashHex, 'utf8');
const secondHashHex = sha256(firstHashBytes);
console.log('  Second hash:', secondHashHex);

const messageHashBytes = Buffer.from(secondHashHex, 'utf8');
console.log('  Message hash (64 bytes):', messageHashBytes.toString('hex'));
console.log('  Expected hash:', testData.messageHash);
console.log('  ✓ Hash matches:', messageHashBytes.toString('hex') === testData.messageHash);
console.log();

// Test 2: Sign the message
console.log('Test 2: BIP-322 Signing');
const privateKey = Buffer.from(testData.privateKey, 'hex');

try {
  const signature = signMessage(privateKey, messageHashBytes);
  console.log('  Generated signature:', signature.toString('hex'));
  console.log('  Expected signature:', testData.signature);
  console.log('  ✓ Signature matches:', signature.toString('hex') === testData.signature);

  // Note: BIP-322 signatures use random nonces, so they will differ each time
  // But the signature should be 64 bytes and should verify correctly
  console.log('  Signature length:', signature.length, 'bytes');

  if (signature.toString('hex') !== testData.signature) {
    console.log('\n  Note: Signatures differ because BIP-322 uses random nonces.');
    console.log('  This is expected - what matters is that verification succeeds.');
  }
} catch (error) {
  console.error('  ✗ Signing failed:', error);
}

File: rust-server/src/test_bip322.rs
// Test program to generate reference BIP-322 signatures for comparison with TypeScript
use arch_program::hash::Hash;
use arch_program::pubkey::Pubkey;
use arch_program::sanitized::{ArchMessage, MessageHeader, SanitizedInstruction};
use arch_sdk::helper::sign_message_bip322;
use bitcoin::key::{Keypair, UntweakedKeypair};
use bitcoin::secp256k1::Secp256k1;
use bitcoin::Network;
use std::str::FromStr;

pub fn test_message_hashing() {
    println!("=== Testing ArchMessage Hashing ===\n");

    // Create a test keypair
    let secp = Secp256k1::new();
    let (secret_key, _public_key) = secp.generate_keypair(&mut rand::thread_rng());
    let keypair = UntweakedKeypair::from_secret_key(&secp, &secret_key);
    let xpubkey = keypair.x_only_public_key().0;
    let pubkey = Pubkey::from_slice(&xpubkey.serialize());

    println!("Test Keypair:");
    println!("  Private key: {}", hex::encode(secret_key.secret_bytes()));
    println!("  Public key: {}", hex::encode(pubkey.serialize()));
    println!();

    // Create a simple test message
    let recent_blockhash = Hash::from([1u8; 32]);
    let program_id = Pubkey::from([2u8; 32]);
    let instruction_data = vec![0x00, 0x01, 0x02, 0x03];

    let message = ArchMessage {
        header: MessageHeader {
            num_required_signatures: 1,
            num_readonly_signed_accounts: 0,
            num_readonly_unsigned_accounts: 0,
        },
        account_keys: vec![pubkey, program_id],
        recent_blockhash,
        instructions: vec![SanitizedInstruction {
            program_id_index: 1,
            accounts: vec![0],
            data: instruction_data.clone(),
        }],
    };

    // Serialize the message
    let serialized = message.serialize();
    println!("Serialized message ({} bytes):", serialized.len());
    println!("  Hex: {}", hex::encode(&serialized));
    println!();

    // Hash the message
    let message_hash = message.hash();
    println!("Message hash ({} bytes):", message_hash.len());
    println!("  Hex: {}", hex::encode(&message_hash));
    println!();

    // Let's manually show the double-hash process
    let first_hash_hex = sha256::digest(&serialized);
    println!("First hash (hex string): {}", first_hash_hex);

    let first_hash_bytes = first_hash_hex.as_bytes();
    println!("First hash as UTF-8 bytes ({} bytes): {}", first_hash_bytes.len(), hex::encode(first_hash_bytes));

    let second_hash_hex = sha256::digest(first_hash_bytes);
    println!("Second hash (hex string): {}", second_hash_hex);

    let second_hash_bytes = second_hash_hex.as_bytes();
    println!("Second hash as UTF-8 bytes ({} bytes): {}", second_hash_bytes.len(), hex::encode(second_hash_bytes));
    println!();

    // Sign the message
    let signature = sign_message_bip322(&keypair, &message_hash, Network::Testnet);
    println!("BIP-322 Signature:");
    println!("  Hex: {}", hex::encode(&signature));
    println!();

    // Verify the signature
    match arch_sdk::verify_message_bip322(
        &message_hash,
        pubkey.serialize(),
        signature,
        false,
        Network::Testnet,
    ) {
        Ok(_) => println!("✓ Signature verification successful!"),
        Err(e) => println!("✗ Signature verification failed: {}", e),
    }
    println!();

    // Output JSON for TypeScript testing
    println!("JSON Test Data:");
    println!("{{");
    println!("  \"privateKey\": \"{}\",", hex::encode(secret_key.secret_bytes()));
    println!("  \"publicKey\": \"{}\",", hex::encode(pubkey.serialize()));
    println!("  \"serializedMessage\": \"{}\",", hex::encode(&serialized));
    println!("  \"messageHash\": \"{}\",", hex::encode(&message_hash));
    println!("  \"signature\": \"{}\"", hex::encode(&signature));
    println!("}}");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn run_test() {
        test_message_hashing();
    }
}
